p2p/host/basic/basic_host_notifiee.go

```diff
@@ -0,0 +1,71 @@
+package basichost
+
+import (
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// basicHostNotifiee listens to Network Notifiee events & emits them on the host's event bus.
+// Please read the docs for the "EvtPeerConnectednessChanged" event in go-libp2p-core for an explanation
+// of the subtleties involved in getting this right.
+type basicHostNotifiee BasicHost
+
+func (b *basicHostNotifiee) basicHost() *BasicHost {
+	return (*BasicHost)(b)
+}
+
+func (b *basicHostNotifiee) Connected(n network.Network, v network.Conn) {
+	p := v.RemotePeer()
+
+	// While the swarm Connected/Disconnected notifications for a single connection are guaranteed
+	// to be serialized, notifications for different connections between the same peer are not
+	// as they are fired concurrently on different go-routines i.e. We could observe a "Connected" notification
+	// for a new connection Conn2 with a peer Px before we observe the "Disconnected" notification for an existing
+	// connection Conn1 with Px even though Conn1 was disconnected before Conn2 was created.
+	// This striped lock ensures that both "Connected" & "Disconnected" notifications synchronize on
+	// "change in connectedness" between two peers & also protects the corresponding `observedConnectedness` state.
+	indexForLk := len(p) - 1
+	stripe := &b.stripedPeerConnectivity[p[indexForLk]]
+	stripe.Lock()
+	defer stripe.Unlock()
+
+	// If our last observed connectedness for the peer was not "Connected" & we are now connected to the peer,
+	// we should emit an event.
+	if stripe.observedConnectedness[p] != network.Connected && n.Connectedness(p) == network.Connected {
+		// change our observed connectedness for the peer to Connected.
+		stripe.observedConnectedness[p] = network.Connected
+
+		// emit the event on the bus
+		evt := event.EvtPeerConnectednessChanged{p, network.Connected}
+		b.basicHost().emitters.evtPeerStateChange.Emit(evt)
+	}
+}
```

Comment:
Don't we want to fire and store iff the current value is _not_ equal to the previous value? If so, we could simplify this comparison, and make it less brittle to actual enum values (what if we introduce new enum values in the future?). This really seems like an atomic compare and swap.

Reply:
We already know that the new state will need to be connected, in this case. We could simplify this but it'll have a minor cost as we'll have to call `n.Connectedness(p)` (takes a lock) _before_ we check the local value.